src/sparseml/exporters/transforms/kv_cache/position_embeddings_adjustment_codegen.py

```diff
@@ -0,0 +1,450 @@
+            [1],
+        )
+        current_sequence_len = onnx.helper.make_tensor_value_info(
+            current_sequence_len_name,
+            TensorProto.INT64,
+            [1],
+        )
+
+        add_node = onnx.helper.make_node(
+            "Add",
+            inputs=[offset_name, current_sequence_len_name],
+            outputs=[total_sequence_len_name],
+            name="add_to_compute_total_sequence_len",
+        )
+
+        graph.add_node(add_node)
+        self.log_match(add_node)
+
+        graph.update()
+        model.graph.input.extend([offset, current_sequence_len])
+
+        return model, offset.name, total_sequence_len_name
+
+    def _update_fixed_pos_embedding(
+        self,
+        model: ModelProto,
+        position_input_name: str,
+        position_embeddings_nodes: List[NodeProto],
+    ) -> ModelProto:
+        """
+        In order to update the ONNX equivalent of:
+        ```
+         sinusoid_inp = (
+                torch.einsum("i , j -> i j",
+                torch.arange(seq_len, dtype=torch.float)
+                inv_freq).to(x.device).float()
+            )
+        ```
+         to
+          sinusoid_inp = (
+                torch.einsum("i , j -> i j",
+                POSITION_EMBEDDINGS,
+                inv_freq).to(x.device).float()
+            )
+        ```
+        we need to transform:
+
+        ```
+        |      ...         ...
+        |       |           |
+        |  Mul (inv_freq) Range
+        |      |          |
+        |       |       |
+        |        |    |
+        |         | |
+        |          |
+        |       Einsum
+        |         |
+        |        ...
+        ```
+        to
+        ```
+        |      ...
+        |       |
+        |  Mul (inv_freq) POSITION_EMBEDDINGS
+        |      |          |
+        |       |       |
+        |        |    |
+        |         | |
+        |          |
+        |       Einsum
+        |         |
+        |        ...
+        ```
+        """
+        graph = ONNXGraph(model)
+        nodes_to_delete = []
+
+        for node in position_embeddings_nodes:
+            # node is the `Range` node
+            child_node = graph.get_node_children(node)[0]
+            # child_node is the `Einsum` node
+            output_to_replace = node.output[0]
+            self.log_match(child_node)
+            for idx, input_name in enumerate(child_node.input):
+                if input_name == output_to_replace:
+                    graph.update_node_input(child_node, position_input_name, idx)
+
+            # remove all the nodes that are not needed anymore
+            # (`Range` node and it's parents)
+            orphaned_nodes = self._find_orphaned_nodes(model, graph, node)
+            nodes_to_delete.extend(orphaned_nodes)
+
+        graph.delete_nodes(nodes_to_delete)
+        graph.update()
+        graph.delete_unused_initializers()
+
+        return model
+
+    def _find_orphaned_nodes(
```

Comment:
can be shared, no?

Reply:
Your previous function traverses the graph top-bottom, mine goes "upstream". We can potentially discuss creating a single helper function.